### PR TITLE
Add key normalisation functionality

### LINKED ISSUE #15: Add key normalisation functionality
## Summary

Add functionality to normalise primary and foreign keys in related tables to zero-based sequential indices while preserving relationships.

## Background

When working with datasets that have non-sequential primary keys (e.g., user_id: 101, 105, 103), it's often useful to normalise these to zero-based sequential indices (0, 1, 2) for consistency and reduced memory usage. This is particularly important when preparing data for machine learning models or creating standardised test datasets.

## Proposed solution

Implement a `normalise_table_keys` function that:
- Takes multiple related tables as input
- Converts primary keys to zero-based sequential integers
- Preserves all foreign key relationships between tables
- Supports both explicit foreign key specification and auto-detection
- Handles edge cases and provides clear error messages

## Use cases

- Preparing data for machine learning models
- Converting legacy database exports
- Standardising key formats across datasets
- Reducing memory usage with compact sequential indices
- Creating consistent test datasets

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
Implements key normalisation functionality to convert primary and foreign keys in related tables to zero-based sequential indices while preserving relationships.

When constructing microdata, we often work with datasets that have inconsistent key formats - some might use sequential IDs starting from 1, others might use large sparse integers like user IDs 101, 105, 103. This creates unnecessary complexity and memory overhead. By normalising all keys to a common zero-based sequential format, we can assume consistent key patterns across all datasets, simplifying downstream processing and reducing memory usage.

The implementation adds `normalise_table_keys()` for multi-table normalisation with relationship preservation and `normalise_single_table_keys()` for single table scenarios. Foreign key relationships are automatically detected based on column name matching, though explicit specification is also supported. The functionality handles edge cases like duplicate keys, missing columns, and invalid references with clear error messages.

Comprehensive test coverage includes 15 test cases covering normal operations, edge cases, and error handling. All existing tests continue to pass and code is formatted with black and isort.

Fixes #15

### CODE REVIEW

File: src/policyengine_data/normalise_keys.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                        break
+
+        if table_fks:
+            foreign_keys[table_name] = table_fks
+
+    return foreign_keys
+
+
+def normalise_single_table_keys(
+    df: pd.DataFrame, key_column: str, start_index: int = 0
```
Comment (juaristi22) [inline src/policyengine_data/normalise_keys.py:152]: key_column:str

no chance we want to pass List[str] in case there's multiple key columns we want to normalize in one table? i guess we can always call the function multiple times then?

File: tests/test_normalise_keys.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        """Test error handling for missing key column."""
+        df = pd.DataFrame({"value": ["A", "B", "C"]})
+
+        with pytest.raises(
+            ValueError, match="Key column 'missing_id' not found"
+        ):
+            normalise_single_table_keys(df, "missing_id")
+
+
+class TestAutoDetectForeignKeys:
```
Comment (juaristi22) [inline tests/test_normalise_keys.py:261]: just out of curiosity, any particular reason why you would choose to organize tests in classes instead of just functions? is it simply for grouping the functionality that each test addresses?

**Review by juaristi22 — APPROVED** [review]
Very nice, very clean. Are we thinking of using this on person or household ids for example?

Small note: I saw you are missing the changelog entry, and I think the reason why that hasn't failed is because there exists a changelog entry in main already (the versioning failed on my code that was last merged).  I was hoping to fix it with my open Dataset migration pr but I think its best if I quickly do it on main already, you might wanna add a a new entry for when thats fixed? Not blocking though (wouldn't be the end of the world if there is no entry when we merge this anyway)
